src/Server.php

```diff
@@ -170,6 +170,12 @@ public function handleRequest(ConnectionInterface $conn, RequestInterface $reque
 
         $this->emit('request', array($request, $response));
 
+        if ($stream instanceof CloseProtectionStream) {
+            $request->emit('end');
+            $request->close();
+            return;
+        }
```

Comment:
Emitting events on another instance should generally be avoided. Also, this looks somewhat similar to the code below, perhaps this can/should be unified?